### ISSUE TITLE
Jakarta EE 10: Support Jakarta Security 3.0

### ISSUE THREAD

**Issue by jhanders34** [OpenLiberty/open-liberty#18741]
## Description
Implement the Jakarta Security 3.0 specification

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Documents

When available, add links to required feature documents. Use "N/A" to mark particular documents which are not required by the feature.

- Aha: Externally raised RFE ([Aha](https://cloud-platform.ideas.aha.io/))
  - Link the RFE with this issue
- UFO: https://ibm.ent.box.com/file/973577297689?s=zrabsaicm2bkfsywft6v6j366u8jeezd 
- FTS: https://github.com/OpenLiberty/open-liberty/issues/24506
- Beta Blog: https://github.com/OpenLiberty/open-liberty/issues/23315
- GA Blog: https://github.com/OpenLiberty/open-liberty/issues/24478

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Process Overview

- [Prioritization](#prioritization)
- [Design](#design)
- [Implementation](#implementation)
- [Legal and Translation](#legal-and-translation)
- [Beta](#beta)
- [GA](#ga)
  - [Focal Point Approvals](#focal-point-approvals-complete-by-feature-complete-date)
- [Other Deliverables](#other-deliverables)

### General Instructions

The process steps occur roughly in the order as presented. Process steps occasionally overlap.

Each process step has a number of tasks which must be completed or must be marked as not applicable ("N/A").

Unless otherwise indicated, the tasks are the responsibility of the Feature Owner or a Delegate of the Feature Owner.

If you need assistance, reach out to the [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect).

**Important: Labels are used to trigger particular steps and must be added as indicated.** 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Prioritization** (Complete Before Development Starts)

The ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) and area leads are responsible for prioritizing the features and determining which features are being actively worked on.

### **Prioritization**
- [x] Feature added to the "New" column of the [Open Liberty project board](https://github.com/orgs/OpenLiberty/projects/2)
  - Epics can be added to the board in one of two ways:
    - From this issue, use the "Projects" section to select the appropriate project board.
    - From the appropriate project board click "Add card" and select your Feature Epic issue
- [x] Priority assigned
  - Attend the Liberty Backlog Prioritization meeting

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Design** (Complete Before Development Starts)

Design preliminaries determine whether a formal design, which will be provided by an [Upcoming Feature Overview (UFO)](https://ibm.box.com/v/UFO-Template) document, must be created and reviewed.  A formal design is required if the feature requires any of the following: UI, Serviceability, SVT, Performance testing, or non-trivial documentation/ID.

### **Design Preliminaries**
- [x] UI requirements identified. (Owner and [UI focal point](https://github.com/orgs/OpenLiberty/teams/ui-approvers))
- [x] ID requirements identified. (Owner and [ID focal point](https://github.com/orgs/OpenLiberty/teams/id-approvers)) : https://github.com/OpenLiberty/docs/issues/5722
   - Refer to [Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty).
   - Feature Owner adds label `ID Required`, if non-trivial documentation needs to be created by the ID team.
   - ID adds label `ID Required - Trivial`, if no design will be performed and only trivial ID updates are needed.
- [x] Serviceability Requirements Identified. (Owner and [Serviceability focal point](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
- [x] SVT Requirements Identified. (Owner and [SVT focal point](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
- [x] Performance testing requirements identified. (Owner and [Performance focal point](https://github.com/orgs/OpenLiberty/teams/performance-approvers))

### **Design**
- [x] POC Design / UFO review requested. (As part of overall Jakarta Security Authentication, Authorization, and Security 3.0 UFO review)
  - Owner adds label `Design Review Request`
- [x] POC Design / UFO review scheduled.
  - Follow the [Adding an Agenda Item](https://github.ibm.com/websphere/POC-Forum#adding-an-agenda-item) instructions
- [x] POC Design / UFO review completed.
- [x] POC / UFO Review follow-ons completed.
- [x] Design / UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) adds label `Design Approved`
  - Add the public link to the UFO in Box to the [Documents](#documents) section.
  - The UFO must always accurately reflect the final implementation of the feature. Any changes must be first approved. Afterwards, update the UFO by creating a copy of the original approved slide(s) at the end of the deck and prepend "OLD" to the title(s). A single updated copy of the slide(s) should take the original's place, and have its title(s) prepended with "UPDATED".

### **No Design**
- [ ] No Design requested.
  - Owner adds label `No Design Approval Request`
- [ ] No Design / No UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - Approver adds label `No Design Approved`

### **FAT Documentation**
- [x] "Feature Test Summary" child task created
  - Use the [Feature Test Summary Template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=)
  - Add FTS issue link to the [Documents](#documents) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Implementation**

A feature must be [prioritized](https://github.com/orgs/OpenLiberty/projects/2) before any implementation work may begin to be delivered (inaccessible/no-ship).  However, a design focused approach should still be applied to features, and developers should think about the feature design prior to writing and delivering any code.  
Besides being prioritized, a feature must also be socialized (or No Design Approved) before any beta code may be delivered.  All new Liberty content must be inaccessible in our GA releases until it is [Feature Complete](#feature-complete) by either marking it `kind=noship` or [beta fencing](#beta-code) it.  
Code may not GA until this feature has obtained the "Design Approved" or "No Design Approved" label, along with all other tasks outlined in the [GA](#ga) section.

### **Feature Development Begins**
- [x] Add the `In Progress` label

## **Legal and Translation**

In order to avoid last minute blockers and significant disruptions to the feature, the **legal items need to be done as early in the feature process as possible**, either in design or as early into the development as possible.  Similarly, translation is to be done concurrently with development.  Both **MUST** be completed before Beta or GA is requested.

### **Legal** (Complete before Feature Complete Date)
- [ ] Changed or new open source libraries are cleared and approved, or N/A. (Legal Release Services/Cass Tucker/Release PM).
- [ ] Licenses and Certificates of Originality (COOs) are updated, or N/A

### **Translation** (Complete 1 week before Feature Complete Date)
- [x] PII updates are merged, or N/A. Note timing with translation shipments.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Beta**

In order to facilitate early feedback from users, all new features and functionality should first be released as part of a beta release.

### **Beta Code**
- [x] Beta fence the functionality
  - `kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`
- [x] Beta development complete and feature ready for inclusion in a beta release
  - Add label `target:beta` and the appropriate `target:YY00X-beta` (where YY00X is the targeted beta version).
- [x] Feature delivered into beta
  - ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager)) adds label `release:YY00X-beta` (where YY00X is the first beta version that included the functionality).

### **Beta Blog** (Complete 1.5 weeks before beta eGA)
- [x] Beta blog issue created and populated using the [Open Liberty BETA blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the beta blog issue in the [Documents](#documents) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **GA**

A feature is ready to GA after it is Feature Complete and has obtained all necessary Focal Point Approvals. 

### **Feature Complete**
- [x] Feature implementation and tests completed.
  - [x] All PRs are merged.
  - [ ] All epic and child issues are closed.
  - [x] All stop ship issues are completed.
- [ ] Legal: all necessary approvals granted.
- [x] Translation: All messages translated or sent for translation for upcoming release 
- [x] GA development complete and feature ready for inclusion in a GA release
  - Add label `target:ga` and the appropriate `target:YY00X` (where YY00X is the targeted GA version).
  - Inclusion in a release requires the completion of all Focal Point Approvals.

### **Focal Point Approvals** (Complete by Feature Complete Date)

These occur only after GA of this feature is requested (by adding a `target:ga` label).  GA of this feature may not occur until all approvals are obtained.

### **All Features**
- [x] **APIs/Externals** Externals have been reviewed or N/A.  ([OpenLiberty/externals-approvers](https://github.com/orgs/OpenLiberty/teams/externals-approvers))
  - Approver adds label `focalApproved:externals` 
- [x] **Demo** Demo is scheduled for an upcoming EOI or N/A. ([OpenLiberty/demo-approvers](https://github.com/orgs/OpenLiberty/teams/demo-approvers))
  - Approver adds label `focalApproved:demo`.
- [x] **FAT** All Tests complete and running successfully in SOE or N/A. ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers))
  - Approver adds label `focalApproved:fat`.
- [x] **Globalization** Translation and TVT are complete or N/A. ([OpenLiberty/globalization-approvers](https://github.com/orgs/OpenLiberty/teams/globalization-approvers))
  - Approver adds label `focalApproved:globalization`.

### **Design Approved Features**
- [x] **Accessibility** Accessibility testing completed or N/A. ([OpenLiberty/accessibility-approvers](https://github.com/orgs/OpenLiberty/teams/accessibility-approvers))
  - Approver adds label `focalApproved:accessibility`.
- [x] **ID** Documentation is complete or N/A. ([OpenLiberty/id-approvers](https://github.com/orgs/OpenLiberty/teams/id-approvers))
  - Approver adds label `focalApproved:id`.
  - > **_NOTE:_**  If only trivial documentation changes are required, you may reach out to the ID Feature Focal to request a `ID Required - Trivial` label.  Unlike features with regular ID requirement, those with `ID Required - Trivial` label do not have a hard requirement for a Design/UFO.
- [x] **Performance** Performance testing is complete or N/A. ([OpenLiberty/performance-approvers](https://github.com/orgs/OpenLiberty/teams/performance-approvers))
  - Approver adds label `focalApproved:performance`.
- [x] **Serviceability** Serviceability has been addressed or N/A. ([OpenLiberty/serviceability-approvers](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
  - Approver adds label `focalApproved:sve`.
- [x] **STE** Skills Transfer Education chart deck is complete or N/A. ([OpenLiberty/ste-approvers](https://github.com/orgs/OpenLiberty/teams/ste-approvers))
  - Approver adds label `focalApproved:ste`.
- [x] **SVT** System Verification Test is complete or N/A. ([OpenLiberty/svt-approvers](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
  - Approver adds label `focalApproved:svt`.

### **Remove Beta Fencing** (Complete by Feature Complete Date)
- [x] Beta guards are removed, or N/A
  - Only after all necessary Focal Point Approvals have been granted.

### **GA Blog** (Complete by Feature Complete Date)
- [x] GA Blog issue created and populated using the [Open Liberty GA release blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the GA Blog issue in the [Documents](#documents) section.

### **Post GA**
- [ ] Replace `target:YY00X` label with the appropriate `release:YY00X`. ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager))

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Other Deliverables**

- [ ] **OL Guides** OL Guides assessment is complete or N/A. ([Yee-Kang Chang](https://github.com/yeekangc))
- [ ] **Standalone Feature Blog Post** A blog post specifically about your feature or N/A.  ([OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect))
  - This should be strongly considered for larger or more prominent features.
  - Follow [instructions](https://github.com/OpenLiberty/blogs/tree/draft#writing-and-publishing-blog-posts-on-the-openlibertyio-blog) in the blogs repo.
- [ ] **WDT** Liberty Developer Tools work is complete or N/A. (Leonard Theivendra)

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

**Comment by teddyjtorres** [comment]
Updated Epic to use newer Open Liberty Feature Template.

**Comment by cthigh** [comment]
UFO review held Monday June 20 covering:
https://github.com/OpenLiberty/open-liberty/issues/18725
https://github.com/OpenLiberty/open-liberty/issues/18726
https://github.com/OpenLiberty/open-liberty/issues/18741 

@garypicher presented.

P 3. (and throughout) The feature should be named appAuthentication rather than jakartaAuthentication 

P 4. (and throughout) The feature should be named appAuthorization rather than jakartaAuthorization  

P 5. appSecurity-4.0 pulls in the public features for  appAuthentication but not appAuthorization since appAuthorization is full profile feature. 

P 7. Are there changes made in preparation for the removal of the Java Security manager? No  What happens when running with Java 18+?  Will authorization fail on 18+?  If the security manager classes are available, then should run OK. Java security doesn’t need to be enabled for authorization to work.  Please confirm if policy checks still work.  Need to test on Java 18. Indicated Java 18 tests are passing.

P 13. Why is appAuthorization-2.1 a point release when methods are being added?

P 14. What happens when both Jakarta security 3.0 and Open ID Connect are enabled together? Can use method from both.

P 26. update feature names jakartaAuthentication > appAuthentication.  jakartaAuthorization > appAuthorization.  OK that names do not align with Jakarta spec versions. 
P 26. Add a note that future work will assist client integration.  #21351

P 27. There will be devEx work for the tools including the Liberty LSP that is behind the spec work.  The tools would validate the API, have auto-completion, etc. No enablement needed, just work needed in the tools. Make a note under the Developer Experience chart. 

P 28. Is this Authentication?  Do we check permissions?  There is a checkPermission method listed but didn’t expect that for Authentication.  It is JASPIC related. Sounds like this is OK. Is this clean up or related to the removal of security manager?

P 32. Test both session database and session cache.

P 34.  Might want to some performance testing since there are new spec implementation.  Contact Jared. 

P 39. What is the migration path for customers already using OIDC?  Will current server configuration be use?  Not until there is some future work.  #21351

**Comment by NottyCode** [comment]
@teddyjtorres The UFO shows the features being `jakartaAuthentication` and `jakartaAuthorization`, but the socialization says that it should be updated to be `appAuthentication` and `appAuthorization`. Has the feedback from the socialization been addressed?

**Comment by teddyjtorres** [comment]
@garypicher Please upload the updated UFO for approval. Thanks!

**Comment by teddyjtorres** [comment]
@NottyCode An updated UFO addressing the feedback was uploaded. Thanks.

**Comment by teddyjtorres** [comment]
Marked feature as beta in https://github.com/OpenLiberty/open-liberty/pull/23008.

**Comment by NottyCode** [comment]
I don't see any response to the comment about 

> UFO review held Monday June 20 covering: #18725 #18726 #18741
> 
> @garypicher presented.
> 
> P 13. Why is appAuthorization-2.1 a point release when methods are being added?

I don't see any response to this question. I suspect the response is trivial, but if we don't address in slides it should be addressed in an issue comment as to why no change is being made.

> 
> P 14. What happens when both Jakarta security 3.0 and Open ID Connect are enabled together? Can use method from both.

Is the update on slide 26 intended to address this question? Slide 14 doesn't and I see slide 26 references something that might be, but I'd like a comment in this issue to indicate how this was resolved and I don't see one.
 
> P 27. There will be devEx work for the tools including the Liberty LSP that is behind the spec work. The tools would validate the API, have auto-completion, etc. No enablement needed, just work needed in the tools. Make a note under the Developer Experience chart.

I would very much prefer more than is provided here. There isn't really enough for the dev tools team to be able to do updates to the Jakarta EE LSP as evidenced by the use of "etc". The point of this isn't to tell the user to use their imagination.

> 
> P 28. Is this Authentication? Do we check permissions? There is a checkPermission method listed but didn’t expect that for Authentication. It is JASPIC related. Sounds like this is OK. Is this clean up or related to the removal of security manager?

Slide 28 doesn't appear to be changed and I'm not clear what the response is to this set of questions.

> P 34. Might want to some performance testing since there are new spec implementation. Contact Jared.

The slides indicate that no change was made here, it isn't clear whether Jared has been contacted and after discussion no performance work is required, of if that just hasn't happened.

**Comment by garypicher** [comment]
> P 13. Why is appAuthorization-2.1 a point release when methods are being added?

The implementations of the updated abstract class and interface are provided by Liberty and consumed by the application, and thus are backwards compatible from an application perspective.  The new methods to get the policy configuration and permissions will be unavailable without recompiling, but existing methods can still be called.  I didn't update the UFO with this information initially because it's background information about why the spec itself was released with a minor version increment; from the perspective of the Liberty implementation, we just copied the version from the spec.

> P 14. What happens when both Jakarta security 3.0 and Open ID Connect are enabled together? Can use method from both.

I removed the text from Slide 26 and elaborated on the answer to this question on Slide 14.  The text on Slide 14 should now fully answer the question, which is that if both are configured for the same application, the annotation will take precedence and the openidConnectClient configuration will be ignored for that request.

> P 27. There will be devEx work for the tools including the Liberty LSP that is behind the spec work. The tools would validate the API, have auto-completion, etc. No enablement needed, just work needed in the tools. Make a note under the Developer Experience chart.

I consulted with Scott Kurz and updated Slide 27 to say "Work with LSP4Jakarta project to identify opportunities for Language Server to assist".  This is not required for completion of the runtime support, but will be useful as a follow-on activity.

> P 28. Is this Authentication? Do we check permissions? There is a checkPermission method listed but didn’t expect that for Authentication. It is JASPIC related. Sounds like this is OK. Is this clean up or related to the removal of security manager?

These changes are to the implementation of AuthConfigFactory provided by the spec, in preparation for the removal of the security manager in a future release.  The checkPermission method shown here is a private method used internally by AuthConfigFactory to check permissions to get/set the factory.  I didn't update the UFO with this information because it relates to internal spec changes and not to the Liberty implementation.

> P 34. Might want to some performance testing since there are new spec implementation. Contact Jared.

Performance testing will not be required for feature sign-off so I have left the slide as n/a.  Per discussion with Jared, the security team will take a follow-on action to provide applications and associated server configurations in the OIDC area for integration into the performance testing environment.

UFO has been updated at --> https://ibm.ent.box.com/file/973577297689

**Comment by teddyjtorres** [comment]
Demo scheduled for the 23.1 EOI.

**Comment by chirp1** [comment]
Info to document received in https://github.com/OpenLiberty/docs/issues/5722. Approving epic.

**Comment by teddyjtorres** [comment]
Hi @OpenLiberty/demo-approvers @tevans78 @cbridgha 
Demo done in the 23.1 EOI on January 17th, 2023.

**Comment by teddyjtorres** [comment]
@OpenLiberty/ste-approvers
STE link: https://ibm.ent.box.com/file/1122190598429

**Comment by teddyjtorres** [comment]
@OpenLiberty/ste-approvers The STE in https://ibm.ent.box.com/file/1122190598429 is complete. Please review and approve if possible. Thanks!

**Comment by teddyjtorres** [comment]
@OpenLiberty/accessibility-approvers - There are no accessibility requirements for this feature. Please let me know approval can be granted or if anything else is needed.

Regards,
Teddy

**Comment by teddyjtorres** [comment]
@OpenLiberty/externals-approvers  - There are only new jakarta APIs from the Jakarta Security specification for this feature. There are no Liberty or REST APIs. Please let me know approval can be granted or if anything else is needed.

Regards,
Teddy

**Comment by teddyjtorres** [comment]
@OpenLiberty/performance-approvers - There are no performance requirements for this feature. Please let me know approval can be granted or if anything else is needed.

Regards,
Teddy

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

**Comment by pgunapal** [comment]
@teddyjtorres Please answer the questions from https://github.com/OpenLiberty/open-liberty/issues/18741#issuecomment-1444456502 for serviceability approval.

**Comment by teddyjtorres** [comment]
@OpenLiberty/serviceability-approvers and @pgunapal - Hi. Here are the answers. Please let me know approval can be granted or if anything else is needed. Thanks!

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

A: Yes

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  

a) What problem paths were tested and demonstrated? A: Configuration problems for the new “@OpenIdAuthenticationMechanismDefinition” annotation were reviewed. Errors performing the Authorization and Token requests. Errors validating the tokens. Errors during Logout and Refresh scenarios. Negative test paths.

b) Who did you demo to? A: The FAT Tester, L3, and other members of Security SSO team, who were not involved in the feature development, reviewed the messages.

c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3? A: Yes, enhancements were suggested to the messages, explanations, and user actions.

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.

a) Who conducted SVT tests for this feature? A: N/A. The feature did not require SVT and SVT approval was already granted.

b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3? A: N/A

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

A: WAS: L3 Security SSO.

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

A: N/A. The feature does not add new metrics or JSON events.

**Comment by dave-waddling** [comment]
FAT Focal approval granted with the proviso that defect 294980 is revisited; see [discussion in Slack](https://ibm-cloud.slack.com/archives/C5TT8KEBE/p1678719791105409).
